.github/workflows/nix.yaml

```diff
@@ -44,46 +52,61 @@ jobs:
           echo "has_backend=${has_backend}" >> $GITHUB_OUTPUT
           echo "Label check result: ${has_backend}"
 
+  # Main build job - runs matrix builds for different architectures
   build:
+    needs: [process-labels]
+    if: |
+      needs.process-labels.outputs.has-backend == 'true' && 
+      (github.event_name != 'pull_request' || 
+      github.event.pull_request.merged == true || 
+      github.event.action != 'closed')
     strategy:
       matrix:
         include:
-          # Direct builds
           - system: x86_64-linux
             host: x86_64-linux
           - system: aarch64-darwin
             host: aarch64-darwin
-          # Cross compilation
           - system: aarch64-linux
             host: aarch64-darwin
-    # needs: [process-labels]
     runs-on: ${{ matrix.host }}
-    # needs.process-labels.outputs.has-backend == 'true' && 
-    #if: |
-    #  (github.event_name != 'pull_request' || github.event.pull_request.merged == true || github.event.action != 'closed')
     outputs:
       docker-image-name: ${{ steps.docker.outputs.image_name }}
+    
     steps:
-      - uses: actions/checkout@v4
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      # Skip aarch64-linux builds if not on main branch
+      - name: Check build conditions
+        id: check-conditions
+        run: |
+          if [[ "${{ matrix.system }}" == "aarch64-linux" && "${{ github.ref }}" != "refs/heads/main" ]]; then
+            echo "should_build=false" >> $GITHUB_OUTPUT
+          else
+            echo "should_build=true" >> $GITHUB_OUTPUT
+          fi
+
       - name: Build all flake outputs
+        if: steps.check-conditions.outputs.should_build == 'true'
         run: |
           om ci run --systems "${{ matrix.system }}" --no-out-link -- -v
 
-      - name: Docker tasks
+      - name: Build Docker image
         id: docker
         if: |
```

Comment:
`should_build` be a condition here as well. Everywhere you use `nix` or `om`, this condition should be checked.

Reply:
With the suggestion below, this would then look like `should_build && should_push_docker`